docs/spec_draft.md

```diff
@@ -304,6 +321,168 @@ the IEEE-754 specification. For boolean element type, the behavior is same as
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.all_gather
+
+### Semantics
+
+For each replica subgroup in `replica_groups`, the operation
+[concatenates](#stablehloconcatenate) the replicas of `input` tensor formed
+using the subgroup along `all_gather_dim` dimension following the order of the
+replicas in the subgroup and produces a `result` tensor.
+
+`channel_handle`, an optional input, is a handle given to a user to represent a
+communication channel between two computations via a
+[stablehlo.send](#stablehlosend) and [stablehlo.recv](#stablehlorecv)
+instruction pair or via collective instructions (e.g.
+[stablehlo.all_reduce](#stablehloall_reduce),
+[stablehlo.all_gather](#stablehloall_gather), etc.). It is represented using a
+`channel-id` and a `channel-type`. A `channel-id` is used for cross-partition
+communication: only operations with the same opcode and `channel-id` can
+communicate to each other. Non-positive `channel-id` is equivalent to no channel
+id. The `channel-type` is not relevant for this instruction.
+
+In the instruction syntax, multiple replica ids are grouped as a replica
+subgroup and multiple replica subgroups form `replica_groups`. The input
+`use_global_device_ids`, if true, interprets the ids in `replica_groups` as
+global device ids given by `rid * P + pid` where P = number of partitions.
+
+In the instruction syntax, multiple replica ids are grouped as a replica
+subgroup and multiple replica subgroups form `replica_groups`. If
+`use_global_device_ids = true`, the ids in `replica_groups` are interpreted as
+global device ids given by `rid * num_pids + pid` where `num_pids` $=$ number of
+partitions.
+
+Given,
+  * `num_pids`: Number of partition ids
+  * `num_rids`: Number of replica ids
+  * `all_partions_ids`: { pid : pid $\in$ [0, `num_pids`) }
+  * `all_replica_ids`: { rid : rid $\in$ [0, `num_rids`) }
+  * `size_subgroup`: size of each subgroup in `replica_groups`.
+  * `num_subgroup`: number of subgroups in `replica_groups`.
+
+The formation of `groups` of execution instances is based on `replica_groups`,
+`channel-id` and `use_global_device_ids` and is defined as follows:
+
+1. If `channel_handle` is not provided and `use_global_device_ids` $=$ `false`
```

Comment:
Let's have consistent `$=$`, `false`, `:`, and `.` at the end of each list here and below.